### PR TITLE
Adding changes to populate the recipe outputs to the resource properties

### PR DESCRIPTION
**Pull request by vishwahiremat**
# Description

Today recipe outputs are added to `outputs[status][binding]` in resource and when access these values in the container we need to use `outputs.status.binding.host`(for host/ property). With this changes adding the output values directly to the resource properties.

- Updated the APIVersions api to get schema details.
- added changed in the dynamic rp processor to filter the recipe output values to check if the values exist in the resource schema
- add changes to directly add values to the resource as part of ApplyDeploymentOutputs
- updated the functional test to valudate the changes.
- added unit test for dynamic rp processor

## Type of change

- This pull request fixes a bug in Radius and has an approved issue (issue link required).
- This pull request adds or changes features of Radius and has an approved issue (issue link required).
- This pull request is a minor refactor, code cleanup, test improvement, or other maintenance task and doesn't change the functionality of Radius (issue link optional).

Fixes: #issue_number

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

- An overview of proposed schema changes is included in a linked GitHub issue.
    - [ ] Yes
    - [ ] Not applicable
- A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
    - [ ] Yes
    - [ ] Not applicable
- The design document has been reviewed and approved by Radius maintainers/approvers.
    - [ ] Yes
    - [ ] Not applicable
- A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
    - [ ] Yes
    - [ ] Not applicable
- A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
    - [ ] Yes
    - [ ] Not applicable
- A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.
    - [ ] Yes
    - [ ] Not applicable

### CODE REVIEW

File: pkg/dynamicrp/backend/processor/dynamicresource.go [hunk 1]
```diff
@@ -65,5 +68,68 @@ func (d *DynamicProcessor) Process(ctx context.Context, resource *datamodel.Dyna
...
+// It retrieves the schema of the resource type and filters out the values that are not part of the schema.
+func addOutputValuestoResourceProperties(ctx context.Context, ucpClient *v20231001preview.ClientFactory, resource *datamodel.DynamicResource, computedValues map[string]any, secretValues map[string]rpv1.SecretValueReference) error {
+
+	ID, err := resources.Parse(resource.ID)
+	if err != nil {
+		return err
+	}
+
+	plane := ID.PlaneNamespace()
+	apiVersionResource, err := ucpClient.NewAPIVersionsClient().Get(ctx, strings.Split(plane, "/")[1], strings.Split(resource.Type, "/")[0], strings.Split(resource.Type, "/")[1], resource.InternalMetadata.UpdatedAPIVersion, nil)
```
Comment (lakshmimsft) [inline pkg/dynamicrp/backend/processor/dynamicresource.go:89]: Should we create variables with comment for what we're passing into the ..NewAPIVersionsClient().Get() call. It will make it more readable.

File: pkg/dynamicrp/backend/processor/dynamicresource.go [hunk 2]
```diff
@@ -65,5 +68,68 @@ func (d *DynamicProcessor) Process(ctx context.Context, resource *datamodel.Dyna
...
+		if contains(resourceProps, key) {
+			resource.Properties[key] = value.Value
+		}
+	}
+
 	return nil
 }
+
+// Helper function to check if a slice contains a specific string
+func contains(slice []string, item string) bool {
```
Comment (lakshmimsft) [inline pkg/dynamicrp/backend/processor/dynamicresource.go:127]: I believe there is a existing function slices.Contains() , if you import golang.org/x/exp/slices
we've used it in our codebase

File: pkg/dynamicrp/backend/processor/dynamicresource_test.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	aztoken "github.com/radius-project/radius/pkg/azure/tokencredentials"
+	"github.com/radius-project/radius/pkg/dynamicrp/datamodel"
+	"github.com/radius-project/radius/pkg/portableresources/processors"
+	"github.com/radius-project/radius/pkg/recipes"
+	"github.com/radius-project/radius/pkg/ucp/api/v20231001preview"
+	"github.com/radius-project/radius/pkg/ucp/api/v20231001preview/fake"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_Process(t *testing.T) {
```
Comment (lakshmimsft) [inline pkg/dynamicrp/backend/processor/dynamicresource_test.go:38]: this is a great test but can we also add a unit test for addOutputValuestoResourceProperties() to handle cases/edge cases which may not be covered here.

File: pkg/dynamicrp/backend/processor/dynamicresource.go [hunk 4]
```diff
@@ -65,5 +68,68 @@ func (d *DynamicProcessor) Process(ctx context.Context, resource *datamodel.Dyna
...
+
+	plane := ID.PlaneNamespace()
+	apiVersionResource, err := ucpClient.NewAPIVersionsClient().Get(ctx, strings.Split(plane, "/")[1], strings.Split(resource.Type, "/")[0], strings.Split(resource.Type, "/")[1], resource.InternalMetadata.UpdatedAPIVersion, nil)
+	if err != nil {
+		return err
+	}
+
+	// Filter out the keyword properties from the resource properties
+	// This is to avoid overwriting the properties like application, environment etc when they are added as computed values or secret values.
+	keywordProperties := []string{"application", "environment", "status"}
```
Comment (lakshmimsft) [inline pkg/dynamicrp/backend/processor/dynamicresource.go:96]: Do we know if this list of keywordProperties is documented outside in some docs/outside of code?

File: pkg/dynamicrp/backend/processor/dynamicresource.go (hunk 1 above)
Comment (vishwahiremat) [reply pkg/dynamicrp/backend/processor/dynamicresource.go:89]: i mean we could if we want, but i think its not necessary during the function call? as function definition already has that information in detail.

File: pkg/dynamicrp/backend/processor/dynamicresource.go (hunk 4 above)
Comment (vishwahiremat) [reply pkg/dynamicrp/backend/processor/dynamicresource.go:96]: its not documented today, but we may want to document this to let the users know that these properties will not be overwritten and if its absolutely necessary users can access them through `status.binding`

File: test/functional-portable/dynamicrp/noncloud/resources/testdata/usertypealpha.yaml [hunk 5]
```diff
@@ -27,21 +27,21 @@ types:
...
               description: The resource ID of the application.
+            database:
+              type: string
+              description: The name of the database.
+            host:
+              type: string
+              description: The host name of the database.
+            port:
+              type: string
+              description: The port number of the database.
```
Comment (lakshmimsft) [inline test/functional-portable/dynamicrp/noncloud/resources/testdata/usertypealpha.yaml:38]: Can we test this with deleting everything after this line

Comment (vishwahiremat) [reply test/functional-portable/dynamicrp/noncloud/resources/testdata/usertypealpha.yaml:38]: yes i have tested that, updated the yaml to the current state as discussed its better to have the test which has accessible fields in resource properties and binding as we support both today

File: pkg/dynamicrp/backend/processor/dynamicresource_test.go (hunk 3 above)
Comment (vishwahiremat) [reply pkg/dynamicrp/backend/processor/dynamicresource_test.go:38]: looks like all the lines in addOutputValuestoResourceProperties are covered but for the err case, i update the code add the error scnario

File: pkg/dynamicrp/backend/processor/dynamicresource.go (hunk 2 above)
Comment (vishwahiremat) [reply pkg/dynamicrp/backend/processor/dynamicresource.go]: thanks for pointing this, updated the code

File: test/functional-portable/dynamicrp/noncloud/resources/testdata/usertypealpha.yaml (hunk 5 above)
Comment (lakshmimsft) [reply test/functional-portable/dynamicrp/noncloud/resources/testdata/usertypealpha.yaml:38]: We do not want the user to specify the fields under status in the yaml file. They will be populated in the backend during recipe processing.

File: pkg/dynamicrp/backend/processor/dynamicresource.go (hunk 4 above)
Comment (lakshmimsft) [reply pkg/dynamicrp/backend/processor/dynamicresource.go]: ccing @Reshrahim, @zachcasper let's make sure this is documented outside of code.

File: pkg/armrpc/asyncoperation/controller/controller.go [hunk 6]
```diff
@@ -81,6 +85,10 @@ func (b *BaseController) DatabaseClient() database.Client {
 	return b.options.DatabaseClient
 }
 
+func (b *BaseController) UcpClient() *v20231001preview.ClientFactory {
```
Comment (lakshmimsft) [inline pkg/armrpc/asyncoperation/controller/controller.go:88]: Comments needed for public function.

**Review by lakshmimsft — APPROVED** [review]
